File: danvas-node/conformance_target.js
// The Node target for the source-SDK conformance suite
// (tests/test_sdk_conformance.py; the normative behavior table lives in
// tests/sdk_conformance_target.py). Run as:
//
//   node danvas-node/conformance_target.js <broker_port>

import { connect } from './index.js'

const port = process.argv[2]
if (!port) {
  console.error('usage: node conformance_target.js <port>')
  process.exit(2)
}

const c = await connect(`127.0.0.1:${port}`, 'ctn')

c.registerTemplate('lbl', 'label', { data: { text: 'hello' }, x: 10, y: 10 })

c.registerTemplate('sld', 'slider',
                   { data: { min: 0, max: 100, value: 10 }, x: 10, y: 110 })
c.onInput('sld', (p) => {
  c.update('lbl', 'post', `v=${Math.round(p.value ?? 0)}`)
})

c.registerTemplate('ask', 'button', { x: 10, y: 210 })
c.onRequest('ask', (data) => ({ pong: ((data && data.ping) || 0) + 1 }))

c.registerTemplate('dl', 'download', { x: 10, y: 310 })
c.onDownload('dl', () => ['hello.txt',
                          new TextEncoder().encode('conformance-bytes\n')])

c.registerTemplate('up', 'upload', { x: 10, y: 410 })
c.onUpload('up', (f) => c.update('lbl', 'post', `up=${f.name}:${f.size}`))

c.register('bin', 'Custom', { html: '<b>bin</b>', w: 240, h: 160 },
           { x: 10, y: 510 })
c.onBinary('bin', (bytes) => {
  c.update('lbl', 'post', `bin=${bytes.length}`)
  c.sendMedia(3, 'bin', bytes) // CUSTOM: echo the bytes back
})

c.registerTemplate('cam', 'video', { x: 10, y: 610 })
c.registerTemplate('ctl', 'button', { x: 10, y: 710 })
c.onInput('ctl', () => {
  const jpeg = new Uint8Array([0xff, 0xd8,
    ...new TextEncoder().encode('conformance-jpeg')])
  c.sendVideo('cam', jpeg)
})

console.log(`[ctn] conformance target live on :${port}`)
// keep the process alive (the socket + heartbeat timer already do)

File: danvas-rust/examples/conformance_target.rs
//! The Rust target for the source-SDK conformance suite
//! (tests/test_sdk_conformance.py; the Python twin and the normative behavior
//! table live in tests/sdk_conformance_target.py).
//!
//!     cargo build --example conformance_target
//!     # then the suite spawns: target/debug/examples/conformance_target <port>

use danvas::Client;
use serde_json::json;

fn main() {
    let port = std::env::args().nth(1).expect("usage: conformance_target <port>");
    let c = Client::connect(&format!("127.0.0.1:{port}"), "ct")
        .expect("connect to danvasd");

    c.label("lbl", "hello").at(10.0, 10.0).show();

    c.slider("sld", 0.0, 100.0, 10.0).at(10.0, 110.0).show();
    let lc = c.clone();
    c.on_input("sld", move |p| {
        let v = p.get("value").and_then(|v| v.as_f64()).unwrap_or(0.0);
        lc.update("lbl", "post", json!(format!("v={v:.0}")));
    });

    c.button("ask").at(10.0, 210.0).show();
    c.on_request("ask", |data| {
        let ping = data.get("ping").and_then(|v| v.as_i64()).unwrap_or(0);
        json!({"pong": ping + 1})
    });

    c.panel("dl", "download").at(10.0, 310.0).show();
    c.on_download("dl", || ("hello.txt".into(), b"conformance-bytes\n".to_vec()));

    c.panel("up", "upload").at(10.0, 410.0).show();
    let uc = c.clone();
    c.on_upload("up", move |f| {
        uc.update("lbl", "post", json!(format!("up={}:{}", f.name, f.size)));
    });

    c.custom("bin", "<b>bin</b>", "", "").at(10.0, 510.0).show();
    let bc = c.clone();
    c.on_binary("bin", move |data| {
        bc.update("lbl", "post", json!(format!("bin={}", data.len())));
        bc.send_media(3, "bin", data); // CUSTOM: echo the bytes back
    });

    c.video("cam").at(10.0, 610.0).show();
    c.button("ctl").at(10.0, 710.0).show();
    let vc = c.clone();
    c.on_input("ctl", move |_| {
        vc.send_video("cam", b"\xff\xd8conformance-jpeg");
    });

    println!("[ct] conformance target live on :{port}");
    std::thread::park();
}
